src/hotspot/share/tsan/tsanOopMap.cpp

```diff
@@ -497,34 +259,37 @@ void TsanOopMap::weak_oops_do(
 // We cannot use ordinary VM mutex, as that requires a state transition.
 void TsanOopMap::add_oop_with_size(oopDesc *addr, int size) {
   DEBUG_ONLY(NoSafepointVerifier nsv;)
-  assert(TsanOopMapImpl::oop_map != NULL, "TsanOopMap not initialized");
+  assert(_oop_map != NULL, "TsanOopMapTable not initialized");
   guarantee(addr != NULL, "null oop");
-  bool alloc = false;
+  bool added = false;
   {
     MutexLocker mu(TsanOopMap_lock, Mutex::_no_safepoint_check_flag);
-    // N.B. addr->size() may not be available yet!
-    alloc = TsanOopMapImpl::oop_map->put(addr, size);
+    added = _oop_map->add_oop_with_size(addr, size);
   }
-  if (alloc) {
+  if (added) {
+    log_trace(tsan)("__tsan_java_alloc for: " PTR_FORMAT ", " PTR_FORMAT "\n",
+                    (uintx)addr, (uintx)addr + size * HeapWordSize);
     __tsan_java_alloc(addr, size * HeapWordSize);
   }
 }
 
 void TsanOopMap::add_oop(oopDesc *addr) {
-  // N.B. oop's size field must be init'ed; else addr->size() crashes.
-  TsanOopMap::add_oop_with_size(addr, addr->size());
+  // We need object size when notify tsan about a freed object.
+  // We cannot call size() for an object after it's freed, so we
+  // need to save the size information in the table.
```

Comment:
> Could we move this comment about the size field to tsanOopMapTable.hpp? It was in tsanOopMap.hpp before.

This is new.

I think you are referring to the following comment in tsanOopMap.hpp:

"// Oop size must be cached, as it is unsafe to call size() after reference is
// collected."

I added following in tsanOopMapTable.hpp:

"// The oop sizes are saved in the table because we need to
// use the size information when notify TSAN about an freed object."